x/gamm/pool-models/stableswap/amm.go

```diff
@@ -276,6 +281,59 @@ func solveCfmmMulti(xReserve, yReserve, wSumSquares, yIn osmomath.BigDec) osmoma
 	return a
 }
 
+// solidly CFMM is xy(x^2 + y^2) = k
+// So we want to solve for a given addition of `b` units of y into the pool,
+// how many units `a` of x do we get out.
+// Let y' = y + b
+// we solve k = (x'y')(x'^2 + y^2) for x', using the following equation {wolfram alpha link}
+// Then we use that to derive the change in x as x_out = x' - x
+func solveCfmmDirect(xReserve, yReserve, yIn osmomath.BigDec) osmomath.BigDec {
+	if !xReserve.IsPositive() || !yReserve.IsPositive() || !yIn.IsPositive() {
+		panic("invalid input: reserves and input must be positive")
+	}
+
+	if yIn.GT(yReserve) {
+		panic("invalid input: cannot trade greater than reserve amount into CFMM")
+	}
+
+	// find k using existing reserves
+	k := cfmmConstant(xReserve, yReserve)
+
+	// find new yReserve after join
+	y_new := yReserve.Add(yIn)
+
+	// store powers to simplify calculations
+	y2 := y_new.Mul(y_new)
+	y3 := y2.Mul(y_new)
+	y4 := y3.Mul(y_new)
+	large_term := (y4.Quo(k)).Mul(osmomath.NewBigDec(2).Quo(twentySevenRootTwo))
+	large_term2 := large_term.Mul(large_term)
+
+	// solve for new xReserve using new yReserve and old k using a solver derived from xy(x^2 + y^2) = k
+	sqrt_term, err := (osmomath.OneDec().Add(large_term2)).ApproxRoot(2)
+	if err != nil {
+		panic(err)
+	}
```

Comment:
Lets take the equation we want to compute, write out the variable substitutions we do, then show that each of this is very clearly computing the desired variable

Reply:
Took a crack at this in the recent commit!